File: Subsets.cpp
#include <vector>
using namespace std;
class Solution {
public:
    vector<int> candidates;
    vector<vector<int>> results;
    vector<vector<int>> subsets(vector<int> &nums) {
        candidates = nums;
        vector<int> empty_vector;
        dfs(0, empty_vector);
        return results;
        
    }
    void dfs(int index, vector<int> path){
        if (path.size() > candidates.size()){
            return;//backtrack
        }
        results.push_back(path);
        for (int i = index; i < candidates.size(); i++){
            vector<int> temp = path;
            temp.push_back(candidates[i]);
            dfs(i+1, temp);
        }   
    }
};

File: Subsets.py
class Solution:
    candidates = []
    results = []
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.candidates = nums
        self.dfs(0,[])
        return self.results

    def dfs(self, index, path):
        if len(path) > len(self.candidates):
            return
        self.results.append(path)
        for i in range(index, len(self.candidates)):
            self.dfs(i+1, path+[self.candidates[i]])
